File: 백준/Dijikstra/1753.js
const fs = require("fs");

const inputs = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
const [v, e] = inputs[0].split(" ").map((e) => Number(e));
const start = Number(inputs[1]);
const nodes = inputs.slice(2).map((e) => e.split(" ").map((e) => Number(e)));

const graph = Array.from({ length: v + 1 }, () => []);
const dist = Array.from({ length: v + 1 }, () => Infinity);

nodes.forEach((node) => {
  const [st, end, weight] = node;
  graph[st].push([end, weight]);
});

class MinHeap {
  constructor() {
    this.heap = [null];
  }

  push(value) {
    this.heap.push(value);
    let current = this.heap.length - 1;
    let parent = Math.floor(current / 2);

    while (parent && this.heap[current][1] < this.heap[parent][1]) {
      [this.heap[current], this.heap[parent]] = [
        this.heap[parent],
        this.heap[current],
      ];
      current = parent;
      parent = Math.floor(current / 2);
    }
  }

  pop() {
    if (this.heap.length === 1) return;
    if (this.heap.length === 2) return this.heap.pop();

    let current = 1;
    let left = 2;
    let right = 3;

    let rv = this.heap[current];

    this.heap[current] = this.heap.pop();

    while (
      (this.heap[left] && this.heap[current][1] > this.heap[left][1]) ||
      (this.heap[right] && this.heap[current][1] > this.heap[right][1])
    ) {
      if (this.heap[right] && this.heap[left][1] > this.heap[right][1]) {
        [this.heap[right], this.heap[current]] = [
          this.heap[current],
          this.heap[right],
        ];
        current = right;
      } else {
        [this.heap[left], this.heap[current]] = [
          this.heap[current],
          this.heap[left],
        ];
        current = left;
      }

      left = current * 2;
      right = left + 1;
    }

    return rv;
  }

  isEmpty() {
    return this.heap.length === 1;
  }
}

const heap = new MinHeap();

heap.push([start, 0]);

dist[start] = 0;

while (!heap.isEmpty()) {
  const [start, weight] = heap.pop();

  graph[start].forEach((v) => {
    const [end, nw] = v;
    if (weight + nw < dist[end]) {
      dist[end] = Math.min(dist[end], weight + nw);
      heap.push([end, weight + nw]);
    }
  });
}

console.log(
  dist
    .slice(1)
    .map((e) => (e === Infinity ? "INF" : e))
    .join("\n")
);

File: 백준/Dijikstra/1753.py
import sys
from heapq import heappush, heappop

vv, e = map(int, sys.stdin.readline().strip().split())
k = int(sys.stdin.readline().strip())

INF = int(1e9)

# Graph representation as adjacency lists (dictionary of dictionaries)
graph = [{} for _ in range(vv + 1)]

# Read in the edges
for _ in range(e):
    u, v, w = map(int, sys.stdin.readline().strip().split())
    if v in graph[u]:
        graph[u][v] = min(graph[u][v], w)
    else:
        graph[u][v] = w

# Distance array initialized to infinity
dists = [INF] * (vv + 1)
dists[k] = 0  # Starting node distance to itself is 0

# Priority queue (min-heap), storing (distance, node)
heap = []
heappush(heap, (0, k))  # Start with the source node (distance 0, node k)

# Dijkstra's algorithm
while heap:
    current_dist, start = heappop(heap)

    # If the current distance is greater than the already found distance, skip it
    if current_dist > dists[start]:
        continue

    # Traverse all adjacent nodes
    for end, weight in graph[start].items():
        new_dist = current_dist + weight

        # If a shorter path is found
        if new_dist < dists[end]:
            dists[end] = new_dist
            heappush(heap, (new_dist, end))

# Output the result
for i in range(1, vv + 1):
    print(dists[i] if dists[i] != INF else "INF")
